source/constructs/api/data_source/service.py

```diff
@@ -409,94 +436,105 @@ def pre_sync(jdbc: SourceJDBCConnection):
     # if not ec2_client.describe_nat_gateways(Filters=[{'Name': 'subnet-id', 'Values': [jdbc.network_subnet_id]}])['NatGateways']:
     #     raise BizException(MessageEnum.SOURCE_SUBNET_NOT_CONTAIN_NAT.get_code(),
     #                        MessageEnum.SOURCE_SUBNET_NOT_CONTAIN_NAT.get_msg())
-    if not ec2_client.describe_availability_zones(Filters=[{'Name': 'zone-name', 'Values': [jdbc.network_availability_zone]}])["AvailabilityZones"]:
+    #  contain gateway && subnet‘s route connected to ngw
+    ngw_list = [item for item in ec2_client.describe_nat_gateways()['NatGateways'] if item['VpcId'] == subnet[0]['VpcId']]
+    if not ngw_list:
+        raise BizException(MessageEnum.SOURCE_VPC_NOT_CONTAIN_NAT.get_code(),
+                           MessageEnum.SOURCE_VPC_NOT_CONTAIN_NAT.get_msg())
+    if not ec2_client.describe_availability_zones(Filters=[{'Name': 'zone-name', 'Values': [connection["AvailabilityZone"]]}])["AvailabilityZones"]:
         raise BizException(MessageEnum.SOURCE_AVAILABILITY_ZONE_NOT_EXISTS.get_code(),
                            MessageEnum.SOURCE_AVAILABILITY_ZONE_NOT_EXISTS.get_msg())
 
-def sync(jdbc: SourceJDBCConnection):
-    logger.info("START SYNC ...")
-    provider_str = gen_jdbc_provider_str(jdbc.account_provider)
-    glue_connection_name = jdbc.instance
-    glue_database_name = f"{provider_str}-{jdbc.instance}-database"
-    crawler_name = f"{provider_str}-{jdbc.instance}-crawler"
-    state = crud.get_jdbc_connection_glue_state(jdbc.account_provider, jdbc.account_id, jdbc.region, jdbc.instance)
-    crawler_role_arn = __gen_role_arn(account_id=_admin_account_id,
-                                      region=_admin_account_region,
-                                      role_name='GlueDetectionJobRole')
-    crud.set_jdbc_connection_glue_state(jdbc.account_provider, jdbc.account_id, jdbc.region, jdbc.instance, ConnectionState.PENDING.value)
-    credentials = gen_credentials(_admin_account_id)
+def sync(glue, lakeformation, credentials, crawler_role_arn, jdbc: JDBCInstanceSourceBase, url: str):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 78 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 99% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.